protocols/kad/src/addresses.rs

```diff
@@ -113,3 +116,39 @@ impl fmt::Debug for Addresses {
         f.debug_list().entries(self.addrs.iter()).finish()
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_remove() {
+        let check_addresses = |expected: &[&Multiaddr], current: &Addresses| {
+            expected.iter().map(|a| *a).collect::<Vec<_>>() == current.iter().collect::<Vec<_>>()
+        };
+
+        let addr1 = "/ip4/127.0.0.1/tcp/1234".parse::<Multiaddr>().unwrap();
+        let addr2 = "/ip4/127.0.0.1/tcp/4321".parse::<Multiaddr>().unwrap();
+        let addr3 = "/ip4/127.0.0.1/tcp/5678".parse::<Multiaddr>().unwrap();
+
+        let mut addresses = Addresses::new(addr1.clone());
+        addresses.insert(addr2.clone());
+        assert!(check_addresses(&[&addr1, &addr2], &addresses));
+
+        // addr3 is not in the list - it should be a no-op and return Ok(())
+        assert_eq!(Ok(()), addresses.remove(&addr3));
+        assert!(check_addresses(&[&addr1, &addr2], &addresses));
+
+        // addr2 is in the list - it should remove it and return Ok(())
+        assert_eq!(Ok(()), addresses.remove(&addr2));
+        assert!(check_addresses(&[&addr1], &addresses));
+
+        // addr2 is not in the list anymore - it should be a no-op and return Ok(()) even if there is only 1 address left
+        assert_eq!(Ok(()), addresses.remove(&addr2));
+        assert!(check_addresses(&[&addr1], &addresses));
+
+        // since addr1 is in the list and it is the only address left, it should be a no-op and return Err(())
+        assert_eq!(Err(()), addresses.remove(&addr1));
+        assert!(check_addresses(&[&addr1], &addresses));
+    }
```

Comment:
I'd prefer to have multiple tests here that look like:

```rust
#[test]
fn given_one_address_when_remove_different_one_returns_ok() {
	let addresses = make_addresses(&[tcp1234()]);

	let res = addresses.remove(tcp5678());

	assert!(res.is_ok());
	assert_eq!(addresses.addrs.len(), 1);
}

fn tcp1234() -> Multiaddr {
	"/ip4/127.0.0.1/tcp/1234".parse().unwrap()
}


fn tcp5678() -> Multiaddr {
	"/ip4/127.0.0.1/tcp/5678".parse().unwrap()
}
```

This follows a pattern of arrange - act - assert and uses a helper functions like `make_addresses` to create new state without much duplication of code.

I think the resulting code is much easier to understand.

Reply:
I have done some changes. Tell me if it does match what you expected.